crates/corecrypto/src/k256_helper.rs

```diff
@@ -0,0 +1,101 @@
+    let secret_key = SecretKey::from_be_bytes(secret_key).map_err(|e| e.to_string())?;
+    let public_key = PublicKey::from_sec1_bytes(public_key).map_err(|e| e.to_string())?;
+    diffie_hellman(&secret_key, &public_key)
+}
+
+/// Verify given a compact signature, recovery_id, digest, and public key in uncompressed format
+/// NOTE: the recovery_id situation is not necessary, but it is a good sanity check
+pub fn verify_sha256(
+    signed_by: &[u8],
+    message: &[u8],
+    signature: &[u8],
+    recovery_id: u8,
+) -> Result<bool, String> {
+    let signing_key = VerifyingKey::from_sec1_bytes(signed_by).map_err(|e| e.to_string())?;
+    let recovery_id = RecoveryId::try_from(recovery_id).map_err(|e| e.to_string())?;
+    let signature = Signature::try_from(signature).map_err(|e| e.to_string())?;
+    // Pre sha256 the message
+    // let digested_message = Sha256::digest(message);
+    let sha256 = Sha256::new().chain(message);
+    let recovered_key = VerifyingKey::recover_from_digest(sha256, &signature, recovery_id)
+        .map_err(|e| e.to_string())?;
+    // Assert that the recovered key matches the signing key
+    if signing_key != recovered_key {
+        return Err("Recovered key does not match signing key".to_string());
+    }
+    signing_key
+        .verify(message, &signature)
+        .map_err(|e| e.to_string())
+        .map(|_| true)
+}
+
+/// Return recovered key from a compact signature, recovery_id, and digest (does sha256 internally)
+pub fn recover_public_key_predigest_sha256(
+    message: &[u8],
+    signature: &[u8],
+) -> Result<Vec<u8>, String> {
+    // Recovery id is the last byte of the signature, split signature into 64:1
+    let (signature, recovery_id) = signature.split_at(64);
+    let recovery_id = recovery_id[0];
+    let recovery_id = RecoveryId::try_from(recovery_id).map_err(|e| e.to_string())?;
+    let signature = Signature::try_from(signature).map_err(|e| e.to_string())?;
+    // Create a pre-digested FixedOutput digest
+    let digest = Sha256::new().chain(message);
+    let recovered_key = VerifyingKey::recover_from_digest(digest, &signature, recovery_id)
+        .map_err(|e| e.to_string())?;
+    Ok(recovered_key.to_encoded_point(false).as_bytes().to_vec())
+}
+
+/// Return recovered key from a compact signature, recovery_id, and message (does keccak256 internally)
+pub fn recover_public_key_predigest_keccak256(
+    message: &[u8],
+    signature: &[u8],
+) -> Result<Vec<u8>, String> {
+    // Recovery id is the last byte of the signature, split signature into 64:1
+    let (signature, recovery_id) = signature.split_at(64);
+    let recovery_id = recovery_id[0];
+    let recovery_id = RecoveryId::try_from(recovery_id).map_err(|e| e.to_string())?;
+    let signature = Signature::try_from(signature).map_err(|e| e.to_string())?;
+    // Create a pre-digested FixedOutput digest
+    let digest = Keccak256::new().chain(message);
```

Comment:
Just out of curiosity, what's the difference between this and the `keccak256` function in `hashes.rs`?

Reply:
Good question! I struggled with this for a while but the recovery method signature cannot take a &[u8] slice. It specifically wants marker traits to denote which Hash function was used. This prevented me from implementing a generic `32 byte scalar + signature => curve math and recovered key` method like iOS wanted originally.


```
error[E0277]: the trait bound `[u8; 32]: FixedOutput` is not satisfied
   --> crates/corecrypto/src/k256_helper.rs:89:59
    |
89  |     let recovered_key = VerifyingKey::recover_from_digest(digest, &signature, recovery_id)
    |                         --------------------------------- ^^^^^^ the trait `FixedOutput` is not implemented for `[u8; 32]`
    |                         |
    |                         required by a bound introduced by this call
    |
    = help: the following other types implement trait `FixedOutput`:
              CoreWrapper<T>
              SimpleHmac<D>
    = note: required for `[u8; 32]` to implement `Digest`
note: required by a bound in `ecdsa::recovery::<impl ecdsa::VerifyingKey<C>>::recover_from_digest`
   --> /Users/michaelx/.cargo/registry/src/github.com-1ecc6299db9ec823/ecdsa-0.15.1/src/recovery.rs:242:12
    |
242 |         D: Digest,
    |            ^^^^^^ required by this bound in `ecdsa::recovery::<impl ecdsa::VerifyingKey<C>>::recover_from_digest`

error[E0277]: the trait bound `[u8; 32]: HashMarker` is not satisfied
   --> crates/corecrypto/src/k256_helper.rs:89:59
    |
89  |     let recovered_key = VerifyingKey::recover_from_digest(digest, &signature, recovery_id)
    |                         --------------------------------- ^^^^^^ the trait `HashMarker` is not implemented for `[u8; 32]`
    |                         |
    |                         required by a bound introduced by this call
    |
    = help: the following other types implement trait `HashMarker`:
              CShake128Core
              CShake256Core
              CoreWrapper<T>
              CtVariableCoreWrapper<T, OutSize, O>
              Keccak224Core
              Keccak256Core
              Keccak256FullCore
              Keccak384Core
            and 12 others
    = note: required for `[u8; 32]` to implement `Digest`
```